session.go

```diff
@@ -582,42 +577,21 @@ func (s *session) getOrCreateStream(id uint16) (*stream, bool) {
 
 // AllowNewStream returns true if a new stream is allowed to be created over
 // this session, and false otherwise.
-func (s *session) AllowNewStream(maxStreamPerConn uint16, idleInterval time.Duration) bool {
+func (s *session) AllowNewStream(maxStreamPerConn uint16) bool {
 	nextID := atomic.LoadUint32(&s.nextID)
 	if nextID > uint32(maxStreamPerConn) {
 		log.Debug("Exhausted maximum allowed IDs on one physical connection, will open new connection")
 		return false
 	}
-	if idleInterval > 0 {
-		now := time.Now()
-		if now.Sub(s.lastDialed) > idleInterval {
-			log.Debugf("No new connections in %v, will start new session", idleInterval)
-			return false
-		}
-	}
 	if s.isClosed() {
 		return false
 	}
 	return true
 }
 
-// MarkDefunct marks this session as defunct. A defunct session will close once
-// all streams are closed.
-func (s *session) MarkDefunct() {
-	s.mx.Lock()
-	s.defunct = true
-	if len(s.streams) == 0 {
-		s.Close()
-	}
-	s.mx.Unlock()
-}
-
 func (s *session) closeStream(id uint16) {
 	delete(s.streams, id)
 	s.closed[id] = true
-	if s.defunct && len(s.streams) == 0 {
-		s.Close()
-	}
```

Comment:
So I tweaked AllowNewStreams to only check if we've reached the maximum number of streams, which will almost never happen in practice. If it does, we'll create a new session, but we don't worry about all the 65K streams on a given session closing and then closing the session because both ever reaching 65K and all 65K closing in a given lantern run are super unlikely, and if it does we just have a dangling TCP connection that will get closed by the server on idle timeout.

In the other case that the current session is closed, we obviously don't have to worry about closing it again.